librz/asm/arch/cil/cil_dis.c

```diff
@@ -0,0 +1,198 @@
+
+#define read_InlineI      read_generic32
+#define read_InlineSig    read_generic32
+#define read_InlineType   read_generic32
+#define read_InlineField  read_generic32
+#define read_InlineString read_generic32
+#define read_InlineTok    read_generic32
+
+#define read_InlineI8 read_generic64
+
+#define read_ShortInlineR read_generic_float
+#define read_InlineR      read_generic_double
+
+// Special
+static int read_InlineMethod(int *pos, CILOp *op, const ut8 *buf, int len) {
+	if (*pos + sizeof(ut32) > len) {
+		return -1;
+	}
+	ut32 tok = rz_read_at_le32(buf, *pos);
+	rz_strbuf_appendf(&op->strbuf, " 0x%X", tok);
+	op->tok = tok;
+	*pos += sizeof(ut32);
+	return 0;
+}
+
+static int read_InlineBrTarget(int *pos, CILOp *op, const ut8 *buf, int len) {
+	if (*pos + sizeof(st32) > len) {
+		return -1;
+	}
+	st32 target = rz_read_at_le32(buf, *pos);
+	rz_strbuf_appendf(&op->strbuf, " %d", target);
+	op->target = target;
+	*pos += sizeof(st32);
+	return 0;
+}
+
+static int read_ShortInlineBrTarget(int *pos, CILOp *op, const ut8 *buf, int len) {
+	if (*pos + sizeof(st8) > len) {
+		return -1;
+	}
+	st8 target = rz_read_at_le8(buf, *pos);
+	rz_strbuf_appendf(&op->strbuf, " %hhd", target);
+	op->target = target;
+	*pos += sizeof(st8);
+	return 0;
+}
+
+static int read_InlineSwitch(int *pos, CILOp *op, const ut8 *buf, int len) {
+	if (*pos + sizeof(ut32) > len) {
+		return -1;
+	}
+
+	ut32 count = rz_read_le32(buf);
+	*pos += sizeof(ut32);
+	if (*pos + count * 4 > len) {
+		return -1;
+	}
+	*pos += count * 4;
+	return 0;
+}
+
+typedef struct {
+	char *str;
+	int (*read_param)(int *pos, CILOp *op, const ut8 *buf, int len);
+	bool prefix;
+} CILOpcodeReader;
+
+/** \file
+ * The opcode_readers arrays include the opcode mnemonic, the
+ * appropriate parameter reader to call, and whether or not the opcode
+ * is a prefix opcode
+ *
+ * eg. OPCODE_SINGLE(CIL_OP_NOP, "nop", InlineNone, 0x00, NEXT)
+ * ->
+ * CILOpcodeReader opcode_readers_single[] = {
+ *   [CIL_OP_NOP] = { .str = "nop", .read_param = read_InlineNone },
+ *   ...
+ * }
+ */
+#define OPCODE_SINGLE(name, string, param, byte, control) [name] = { .str = string, .read_param = read_##param },
+static const CILOpcodeReader opcode_readers_single[] = {
+#include "opcodes_single.def"
+	[0xFF] = { 0 }
+};
+
+#define OPCODE_DOUBLE(name, string, param, byte, control) [name] = { .str = string, .read_param = read_##param },
+#define OPCODE_PREFIX(name, string, param, byte, control) [name] = { \
+	.prefix = true, \
+	.str = string, \
+	.read_param = read_##param \
+},
+static const CILOpcodeReader opcode_readers_double[] = {
+#include "opcodes_double.def"
+#include "opcodes_prefix.def"
+	[0xFF] = { 0 }
+};
+
+/**
+ * \brief Disassemble a CIL buffer
+ * \return 0 on success, -1 on fail
+ */
+int cil_dis(CILOp *op, const ut8 *buf, int len) {
+	int pos = 0;
+	if (pos >= len) { // pos + 1 > len
+		return -1;
+	}
+
+	ut8 byte;
+
+start: // Taken after a prefix opcode has been consumed
+	byte = buf[pos++];
```

Comment:
i'm not a big fan of gotos used to create loops, but i guess is acceptable since the code is very readable.